text/0712-query-params-as-derived-data.md

```diff
@@ -0,0 +1,153 @@
+---
+Stage: Accepted
+Start Date: 2020-01-26
+Release Date: Unreleased
+Release Versions:
+  ember-source: vX.Y.Z
+  ember-data: vX.Y.Z
+Relevant Team(s): Ember.js, Learning
+RFC PR: https://github.com/emberjs/rfcs/pull/712
+---
+
+<!---
+Directions for above:
+
+Stage: Leave as is
+Start Date: Fill in with today's date, YYYY-MM-DD
+Release Date: Leave as is
+Release Versions: Leave as is
+Relevant Team(s): Fill this in with the [team(s)](README.md#relevant-teams) to which this RFC applies
+RFC PR: Fill this in with the URL for the Proposal RFC PR
+-->
+
+# Query Params as derived data
+
+## Summary
+
+Query Params are awkward in ember in that they align more to the paradigms of a much older
+ember. In the spirit of Octane, this RFC proposes to update how we think about query params
+such that they align with the idea that they are derived state from the URL, much like
+native getters have become in components / services / etc
+
+## Motivation
+
+As taught in [the current version of the guides (3.24)](https://guides.emberjs.com/v3.24.0/routing/query-params/),
+
+Query Params do not currently follow the spirit of Octane:
+ - All data is derived from some source of data
+ - Unidirectional data flow, eliminating "spooky action from a distance"
+
+> All data is derived from some source of data
+
+The source of the data when it comes to Query Params is the URL. With the current way
+Query Params are taught, there are two sources: the URL, and the (sometimes @tracked) property
+on the controller.
+
+```js
+export default class ArticlesController extends Controller {
+  queryParams = ['category'];
+
+  // looks like it could be the source of data, but is also "spookily" overwritten during transitions
+  category = null;
+}
+```
+
+> Unidirectional data flow, eliminating "spooky action from a distance"
+
+There are several ways to update both a query param and the URL.
+With the above example, `category` is _two way bound_ to the URL -- the URL updates `category`,
+and updates to `category` cause transitions to the URL.
+
+Explicit transitions can cause the URL to change, which then cause `category` to change as well.
+
+
+**Proposal**: we teach a more direct way of interacting with query params via "derived data".
+
+Example,
+
+```js
+import Controller from '@ember/controller';
+import { action } from '@ember/object';
+import { inject as service } from '@ember/service';
+
+export default class ArticlesController extends Controller {
+  @service router;
+
+  queryParams = ['category'];
```

Comment:
I like this direction. Do you consider how query params are declared in scope for this RFC?

Reply:
Maybe discussing them is -- as it's def part of the learning story.

I also don't think query params should have default values (especially since that state is entangled with the controller). 
For the URL to be the source of truth, to access a QP from within a component would require something like this:

```ts
get category() {
     return this.router.currentRoute.queryParams.category ?? 'default value';
}
```

Reply:
related: https://github.com/emberjs/rfcs/pull/715